[stablehlo] Add attribute ceil_mode to reduce_window op

sshonTT opened the pull request:
In stablehlo, pool layer converts to reduce_window op, but it has no explicit field for `ceil_mode`.

brataTT commented on lib/Conversion/TorchToStablehlo/Pooling.cpp:
```diff
@@ -520,6 +521,8 @@ class ConvertAtenMaxPoolOp : public ConvertAtenOp<AtenOpT> {
           const int64_t extraPadding = sizeDiff * stride[i];
           stablehloPadding[frontPadIdx] += extraPadding / 2;
           stablehloPadding[backPadIdx] += extraPadding - extraPadding / 2;
```
Thanks for recreating the PR to target the dev branch and taking care of unit tests.
Copying comments from previous PR here to continue the conversation.

> Would it make sense to get rid of previous added padding and only keep ceilmode padding here?
Or do we want to keep both, then handle it in ttir?

> And to verify the change, are we testing against vovnet in tt-torch?
If I remember right, that was one of the models that needed ceilmode.

sshonTT replied:
I intend to keep both because there may be user-specified padding, so I'd like to preserve both and handle them appropriately in TTIR.

And yes, you're right about vovnet, with this change (and the corresponding TTIR update), vovnet achieves good PCC.

sshonTT committed:
[stablehlo] Add attribute ceil_mode to reduce_window op

In stablehlo, pool layer converts to reduce_window op, but it has no explicit field for `ceil_mode`.

mmanzoorTT commented on lib/Conversion/TorchToStablehlo/Pooling.cpp:
```diff
@@ -520,6 +521,8 @@ class ConvertAtenMaxPoolOp : public ConvertAtenOp<AtenOpT> {
           stablehloPadding[frontPadIdx] += extraPadding / 2;
           stablehloPadding[backPadIdx] += extraPadding - extraPadding / 2;
+          ceilModePadding[frontPadIdx] += extraPadding / 2;
+          ceilModePadding[backPadIdx] += extraPadding - extraPadding / 2;
```
`ceilModePadding` is computed (it is duplicate of `stablehloPadding`) but never used. You may remove `ceilModePadding` completely.

sshonTT replied:
It is used in tt-mlir, stablhlo to ttir conversion to calculate original padding size given by user

mmanzoorTT commented on lib/Conversion/TorchToStablehlo/Pooling.cpp:
```diff
@@ -539,6 +542,19 @@ class ConvertAtenMaxPoolOp : public ConvertAtenOp<AtenOpT> {
+    // earlier.
+    mlir::BoolAttr ceilModeAttr = rewriter.getBoolAttr(ceilMode);
+    reduceWindowOp->setAttr(llvm::StringRef("ceil_mode"), ceilModeAttr);
+    DenseIntElementsAttr ceilModePad = DenseIntElementsAttr::get(
```
It seems that `pad` and `ceilModePad` are referring to same padding. Why not only add boolean `ceil_mode` attribute and if `ceil_mode` is true then use `pad`. We don't need to add same paddings twice with two different names.

sshonTT replied:
ceilModePad only contains the padding size that is calculated when ceil_mode is set to true.
So if the user has explicitly specified padding values, they are not included here.

mmanzoorTT reviewed (commented):
I'll suggest to submit this PR in upstream repo first and let's see what others think about it.

mmanzoorTT commented on lib/Conversion/TorchToStablehlo/Pooling.cpp:
```diff
@@ -539,6 +542,19 @@ class ConvertAtenMaxPoolOp : public ConvertAtenOp<AtenOpT> {
+        RankedTensorType::get(
+            {static_cast<int64_t>(inputRank), static_cast<int64_t>(2)},
+            rewriter.getI64Type()),
+        stablehloPadding);
```
My comments/confusion is due to this line. Do you want to add `ceilModePadding` or `stablehloPadding`?
```suggestion
        ceilModePadding);
```

sshonTT replied:
Oh silly me. It is definitely ceilModePadding. Thank you!

sshonTT committed:
Update lib/Conversion/TorchToStablehlo/Pooling.cpp

Co-authored-by: Muhammad Asif Manzoor <mmanzoor@tenstorrent.com>
